Added a normalized offset parameter for all sprites

TwistedTwigleg opened the pull request:
Added a normalized offset parameter for all sprites, including all of the 3D sprites, named "anchor".
Closes #9490

I'm not totally sure the code for Sprite3D and AnimatedSprite3D is working, because I wasn't able to make a visible Sprite3D or AnimatedSprite3D for testing. I think the code works, but someone should probably look it over just to be sure.

Another thing I'm not sure about is having a use_anchor property for Sprite2D andAnimatedSprite2D. I had to included it for Sprite3D, which makes me think I probably should include it in Sprite2D and AnimatedSprite2D. What do you guys think?

TwistedTwigleg committed:
Added an anchor parameter to all sprites (including sprite 3D)
Anchor is a normalized vector that sets offset to the sprite's image bounds.

bojidar-bg commented on scene/2d/animated_sprite.cpp:
```diff
@@ -490,6 +490,34 @@ Point2 AnimatedSprite::get_offset() const {
+
+	anchor = p_anchor;
+	anchor.x = CLAMP(anchor.x, -1, 1);
+	anchor.y = CLAMP(anchor.y, -1, 1);
```
I don't think you should clamp the anchor, actually I think there might be a few uses where an unclamped anchor makes sense.

TwistedTwigleg replied:
I'll remove the clamp. Out of curiosity, what would be a use case for an unclamped anchor?
(I couldn't think of any, which is why I had the value clamped)

bojidar-bg commented on scene/2d/animated_sprite.cpp:
```diff
@@ -490,6 +490,34 @@ Point2 AnimatedSprite::get_offset() const {
+
+	if (centered) {
+		calc_anchor.x = CLAMP(anchor.x, -0.5, 0.5);
+		calc_anchor.y = CLAMP(anchor.y, -0.5, 0.5);
```
Don't use just want to add (0.5,0.5)?

TwistedTwigleg replied:
No, I wanted to clamp the values in range (-0.5 to 0.5) so if a sprite is centered so it would stay at the edges and not go further.

If I remove clamping, should I do some calculations here so that the position stays the same if the sprite is centered or not? For example, a anchor of (0, 1) would position the node at the sprite's feet when not centered. If centered though, you would have to use (0, 0.5) to get that same position. Should I detect if a sprite is centered and make adjustments as necessary to keep the positioning the same?

bojidar-bg commented on scene/2d/animated_sprite.cpp:
```diff
@@ -490,6 +490,34 @@ Point2 AnimatedSprite::get_offset() const {
+	Ref<Texture> texture = frames->get_frame(animation, frame);
+	s = Size2(texture->get_size());
+
+	offset = Point2(s.width * calc_anchor.x, s.height * -calc_anchor.y);
```
Changing offset.. I think you should just calculate it in NOTIFICATION_DRAW, as someone might want to use _both_ anchor and offset. (Also, consider what happens when the size of the texture changes. You also have to update the offset in this case, otherwise it is useless).

TwistedTwigleg replied:
Makes sense. Should I add offset to the calculated anchor value? So it's anchor/normalized-offset + offset?

bojidar-bg commented on scene/2d/sprite.cpp:
```diff
@@ -164,6 +164,36 @@ Point2 Sprite::get_offset() const {
+	update();
+	item_rect_changed();
+	_change_notify("anchor");
+}
```
The three comments from above apply here too.

bojidar-bg commented on scene/3d/sprite_3d.cpp:
```diff
@@ -360,6 +383,23 @@ void Sprite3D::_draw() {
+		else
+			tex_size = Size2(texture->get_size()) / Size2(hframes, vframes);
+
+		offset = Point2(tex_size.width * calc_anchor.x, tex_size.height * -calc_anchor.y);
```
Cool, but you want to be able to use both offset and anchor. (Note that you are actually overriding the offset here, which is bad, as you don't even notify the editor about it)

TwistedTwigleg replied:
Yeah, I should have notified the editor about the changes in offset... Whoops :cold_sweat: 

On a slight aside, do 3D sprites work for you? I wasn't able to get them to work, so I couldn't test if the anchor code was working.

bojidar-bg reviewed (commented):
Additionally, consider removing the use_anchor property -- if `anchor` is (0, 0), then it would make no change at all.

TwistedTwigleg commented:
I changed the code based on the review:
* use_anchor and it's functions in Sprite3D and AnimatedSprite3D have been removed
* Anchor now effects the calculated offset (ofs) in the drawing code for each sprite, not the user inputted offset
* All anchor clamps have been removed

I added a condition to check if the sprite is centered or not and make adjustments accordingly, but it turned out to be more confusing than just having to make adjustments manually, so I removed it.

The code is now both smaller, and more concise. A win-win combo! :smiley:

reduz commented:
I'm not very convinced this is needed. Did you know you can press "v" where the mouse pointer is and it will change the pivot to that? If you want to set it to a specific exact place like bottom, left, center, etc we could add a dialog like "Edit -> Set Pivot To.." 
https://snag.gy/WjNqyO.jpg

TwistedTwigleg commented:
I opened this pull request to close #9490, where it was suggested to add an anchor property. The idea is to have the ability to easily set the offset to the edges of a sprite.

I did not know you can press "v" and it will set the offset to the mouse pointer, good to know. It even works with snapping, including pixel snapping, which might make this PR pointless.

However, I can see one potential reason you might want an anchor property. If you have a sprite that is changing sizes and you want the offset to always be at the sprite's edge or at some offset based on the texture's size.

For example, say you have a character sprite, let's say classic Mario. Say you want to have the offset at the bottom of the sprite, his feet. When he get's a powerup he becomes taller (goes from one tile high to two tiles high). Using anchor you would still have the pivot set at the sprite's feet despite the size of the sprite since it's calculated at runtime.

I'm totally okay with closing this PR, making an editor dialog, or something else. It's all okay with me, I'm just explaining why I think it could be useful 😀

(I think we should add a editor dialog if we close this PR, since there are certain cases where I could see having the ability to set the pivot to the sprite's texture bounds useful and it would still close #9490)

reduz commented:
Closing. The case you describe, while it can indeed be useful, seems like a rare situation that might not happen often. As such, having an extra property just for this is likely too much.

Still, thanks for the work spent on the PR!, It also seems that we should find a way to make the "v" function more obvious, so any feedback on that is very welcome!

TwistedTwigleg commented:
I wasn't sure if the PR would be useful after hearing about the "v" shortcut, so I think closing the PR is a good idea. 😃 

As far as the shortcut goes, maybe a list? Perhaps adding an option in the editor that shows all of the shortcuts and their default values? This might help with other shortcuts, like the shortcut suggested in #9389 for example.

akien-mga commented:
You mean like this? :)
![spectacle wx6278](https://user-images.githubusercontent.com/4701338/28292770-e7952d6a-6b50-11e7-8173-b13967405fc0.png)

("V" doesn't appear there though)

akien-mga commented:
It's defined here: https://github.com/godotengine/godot/blob/34c2024/editor/plugins/canvas_item_editor_plugin.cpp#L236

It should be modified to use a proper shortcut like https://github.com/godotengine/godot/blob/34c2024/editor/plugins/canvas_item_editor_plugin.cpp#L3331, but all those seem linked to Controls, I'm not sure how "global" shortcuts are handled.

TwistedTwigleg commented:
Yeah, like that 😅 

I can try to convert "V" to a proper shortcut later if you guys want. I won't be able to work on it for awhile though.
